### PR TITLE
feat(rome_js_formatter): Binary parentheses

### PR DESCRIPTION
**Pull request by MichaReiser**
This PR refines the rules for setting parentheses around binary expression to more closely match Prettier's behaviour.

It does so by re-using `OperatorPrecedence` enum from the parser crate (which is why it is moved to `rome_js_syntax`).

## Tests

Reviewed the updates to the prettier snapshots.

**File Based Average Prettier Similarity**: 78.58%  -> 78.85%
**Line Based Average Prettier Similarity**: 74.00%   -> 74.25%

### CODE REVIEW

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 1]
```diff
@@ -177,41 +200,41 @@ pub(crate) enum BinaryLikeOperator {
...
+            JsAnyBinaryLikeLeftExpression::JsAnyExpression(JsAnyExpression::JsLogicalExpression(_))
+        )
+    {
+        return Ok(parent_operator != current_operator);
+    }
+
+    let current_precedence = current_operator.precedence();
+    let parent_precedence = parent_operator.precedence();
+
+    #[allow(clippy::if_same_then_else)]
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:224]: I suppose you added this because it helps for readability?

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 2]
```diff
@@ -831,30 +854,46 @@ impl JsAnyBinaryLikeExpression {
...
-                    instance_of.left()?,
-                    JsAnyExpression::JsInstanceofExpression(_)
-                )
-            }
-            JsAnyBinaryLikeExpression::JsInExpression(in_expression) => matches!(
-                in_expression.property()?,
-                JsAnyInProperty::JsAnyExpression(JsAnyExpression::JsInExpression(_))
-            ),
-        })
+        let left = self.left()?;
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:857]: Is the comment still valid? Should it be updated?

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs [hunk 3]
```diff
@@ -831,30 +854,46 @@ impl JsAnyBinaryLikeExpression {
...
+    } else if parent_precedence.is_equality() && child_precedence.is_equality() {
+        false
+    } else if (matches!(
+        child_operator,
+        BinaryLikeOperator::Binary(JsBinaryOperator::Remainder)
+    ) && parent_precedence.is_multiplicative())
+        || (matches!(
+            parent_operator,
+            BinaryLikeOperator::Binary(JsBinaryOperator::Remainder)
+        ) && child_precedence.is_multiplicative())
```
Comment (ematipico) [inline crates/rome_js_formatter/src/utils/binary_like_expression.rs:887]: The formatting here is really hard to follow for me 🤣 I wish `rustfmt` was at prettier level here. Just a rant, carry on!

**Review by ematipico — APPROVED** [review]
Great progresses! 🎆

File: crates/rome_js_formatter/src/utils/binary_like_expression.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/rome_js_formatter/src/utils/binary_like_expression.rs:224]: Yes, I find it easier to reason if every `if` handles a specific case rather than grouping them all together because the main point here isn't about the body (which is a simple bool) but about the condition.
